projects/ad2s1210_iio/app/ad2s1210_iio.c

```diff
@@ -0,0 +1,619 @@
+static int8_t data_buffer[DATA_BUFFER_SIZE] = { 0 };
+#endif
+
+#define swab16(x) \
+        ((((x) & 0x00ff) << 8) | \
+         (((x) & 0xff00) >> 8))
+
+#define AD2S1210_ENABLE_HYSTERESIS	NO_OS_BIT(4)
+
+enum ad2s1210_attribute_id {
+	RAW_ATTR_ID,
+	SCALE_ATTR_ID,
+	SAMPLING_FREQ_ATTR_ID,
+	LABEL_ATTR_ID,
+	HYSTERESIS_ATTR_ID,
+	HYSTERESIS_AVAILABLE_ATTR_ID,
+	FREQ_ATTR_ID,
+	FREQ_AVAIL_ATTR_ID,
+};
+
+/******************************************************************************/
+/******************** Variables and User Defined Data Types *******************/
+/******************************************************************************/
+
+/* Pointer to the struct representing the AD2S1210 IIO device */
+struct ad2s1210_dev *ad2s1210_dev_inst;
+
+/* IIO interface descriptor */
+static struct iio_desc *ad2s1210_iio_desc;
+
+#if (DATA_CAPTURE_MODE == CONTINUOUS_DATA_CAPTURE)
+static struct iio_hw_trig *ad2s1210_hw_trig_desc;
+#endif
+
+struct scan_type chn_scan[RESOLVER_CHANNELS] = {
+	{
+		.sign = 'u',
+		.realbits = CHN_STORAGE_BITS,
+		.storagebits = CHN_STORAGE_BITS,
+		.shift = 0,
+		.is_big_endian = true,
+	},
+	{
+		.sign = 's',
+		.realbits = CHN_STORAGE_BITS,
+		.storagebits = CHN_STORAGE_BITS,
+		.shift = 0,
+		.is_big_endian = true,
+	},
+	{
+		.sign = 'u',
+		.realbits = CHN_STORAGE_BITS,
+		.storagebits = CHN_STORAGE_BITS,
+		.shift = 0,
+		.is_big_endian = false,
+	},
+};
+
+/* IIO channels attributes list */
+static struct iio_attribute
+	ad2s1210_iio_ch_attributes[RESOLVER_CHANNELS][RESOLVER_MAX_ATTR] = {
+	{
+		AD2S1210_CHN_ATTR("hysteresis", HYSTERESIS_ATTR_ID),
+		AD2S1210_CHN_ATTR("hysteresis_available", HYSTERESIS_AVAILABLE_ATTR_ID),
+		AD2S1210_CHN_ATTR("label", LABEL_ATTR_ID),
+		AD2S1210_CHN_ATTR("raw", RAW_ATTR_ID),
+		AD2S1210_CHN_ATTR("scale", SCALE_ATTR_ID),
+		END_ATTRIBUTES_ARRAY
+	},
+	{
+		AD2S1210_CHN_ATTR("label", LABEL_ATTR_ID),
+		AD2S1210_CHN_ATTR("raw", RAW_ATTR_ID),
+		AD2S1210_CHN_ATTR("scale", SCALE_ATTR_ID),
+		END_ATTRIBUTES_ARRAY
+	},
+	{
+		AD2S1210_CHN_ATTR("frequency", FREQ_ATTR_ID),
+		AD2S1210_CHN_ATTR("frequency_available", FREQ_AVAIL_ATTR_ID),
+		AD2S1210_CHN_ATTR("label", LABEL_ATTR_ID),
+		END_ATTRIBUTES_ARRAY
+	},
+};
+
+/* IIO device (global) attributes list */
+static struct iio_attribute ad2s1210_iio_global_attributes[] = {
+	AD2S1210_CHN_ATTR("sampling_frequency", SAMPLING_FREQ_ATTR_ID),
+	END_ATTRIBUTES_ARRAY
+};
+
+/* IIO channels info */
+static struct iio_channel ad2s1210_iio_channels[RESOLVER_CHANNELS] = {
+	AD2S1210_CH("position", 0, IIO_ANGL, false),
+	AD2S1210_CH("velocity", 1, IIO_ANGL_VEL, false),
+	AD2S1210_CH("altvoltage", 2, IIO_ALTVOLTAGE, true)
+};
+
+/* Flag to indicate if size of the buffer is updated according to requested
+ * number of samples for the multi-channel IIO buffer data alignment
+ */
+static bool buf_size_updated;
+
+static const float ad2s1210_velocity_scale[] = {
+	17089132, /* 8.192MHz / (2*pi * 2500 / 2^15) */
```

Comment:
Can we add macros defined with the equation mentioned in the comment for these values?

Reply:
ok